vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java

```diff
@@ -84,35 +96,39 @@
 @JsModule("./comboBoxConnector.js")
 public class ComboBox<T> extends GeneratedVaadinComboBox<ComboBox<T>, T>
         implements HasSize, HasValidation,
-        HasFilterableDataProvider<T, String> {
+        HasDataView<T, String, ComboBoxDataView<T>>,
+        HasListDataView<T, ComboBoxListDataView<T>>,
+        HasLazyDataView<T, String, ComboBoxLazyDataView<T>> {
 
     private static final String PROP_INPUT_ELEMENT_VALUE = "_inputElementValue";
     private static final String PROP_SELECTED_ITEM = "selectedItem";
     private static final String PROP_VALUE = "value";
 
     /**
      * A callback method for fetching items. The callback is provided with a
-     * non-null string filter, offset index and limit.
+     * non-null filter, offset index and limit.
      *
      * @param <T>
      *            item (bean) type in ComboBox
+     * @param <F>
+     *            filter type
      */
     @FunctionalInterface
-    public interface FetchItemsCallback<T> extends Serializable {
+    public interface FetchItemsCallback<T, F> extends Serializable {
```

Comment:
> So existing usage of String for the filter type should compile and get the same behavior, right ?
Yes, it should. From the client code perspective, it should be fine. They can use `void setDataProvider(FetchItemsCallback<T, String> fetchItems, SerializableFunction<String, Integer> sizeCallback)` as usual. From binary compatibility, it should be also fine, because `F` becomes `Object` after erasure, and it can obtain `String`.

> Another issue might be for the component which extends and reuses this
Yes, that is inconvenience. But from the other side. We have those two methods clashing:

- `setItems(CallbackDataProvider.FetchCallback<T, C> fetchCallback, SerializableFunction<String, C> filterConverter)`
- `setItems(CallbackDataProvider.FetchCallback<T, C> fetchCallback, CallbackDataProvider.CountCallback<T, C> countCallback`)

Compiler can't decide which one to use for this, because `can` be interpreted as a `Integer`:

```
        comboBox.setItems(query -> service.fetch(...), query -> 42);
```

Even swapping the arguments won't help. I see no other elegant solution for that so far.

Reply:
It has been too long time since the initial review - did we sort this out somehow ?
Why did we choose to have `F` type instead of `String` which the CB always provides and then force the conversion to happen in the callback implementation instead ?

Reply:
These changes have been reverted and `setItems` methods which take the filter converter renamed to `setItemsWithFilterConverter`.